### DISCUSSION TITLE
How do you package a font into p_fa.mpc?

### DISCUSSION THREAD

**Discussion by andy013** [Albeoris/Memoria#995] [Q&A] [answered]
Hi, I searched around on here and discord but couldn't find any info on how to package a font into the p_fa.mpc file. Do you need to use unity for it or did someone write some code to do it? Also, is it possible to extract fonts from existing p_fa.mpc files?

Thanks

**Comment by andy013** [comment, marked as answer]
I spent a little while playing around with this and I figured out how to bundle a font into p_fa.mpc.

**Note: This is for developers who want to ship a font with Memoria.** If you just want to use a custom font in-game for yourself then follow the [steps below](https://github.com/Albeoris/Memoria/discussions/995#discussioncomment-11710038) by @Tirlititi.


In order to import a font, it must be inside a Unity asset bundle. When you use Unity to generate the font asset it automatically generates a texture and material that are needed too. I couldn't figure out any way to import a font directly and get these to generate during runtime. You can use `Font.CreateDynamicFontFromOSFont()` but this only works for fonts already installed on the users system. 

The `p_fa.mpc` file is just an asset bundle that has been encrypted. Memoria includes a `ByteEncryption` class that can be used to encrypt and decrypt this file. 


Here are the steps for bundling a custom font:

1. Install Unity Hub from here: https://unity.com/download

2. Download Unity version 5.2.3p2. This is the version that FF9 runs on. You can find it here: https://unity.com/releases/editor/archive by clicking the Unity 5 tab and then the All Versions tab. 

3. Start a new project, making sure to select editor version 5.2.3p2 and open it with the editor

![1](https://github.com/user-attachments/assets/26876fcc-a433-4221-b1ad-ac8cc7d27f52)
*I'm using 5.2.3f1 here. It works fine too.*

4. Drag and drop your font into the Assets folder within the asset browser at the bottom of the editor window. This will generate a font asset that includes a material and a texture.

![2](https://github.com/user-attachments/assets/c8b8cbc4-9f87-469c-8d81-793b8fe3bd07)

5. Create a C# script by right clicking on the asset browser and going to the Create menu. Call it `FontBundler` and double click it to open it.

![3](https://github.com/user-attachments/assets/170346cd-1205-4e06-b86d-03d548a879c7)

6. Paste this code into the script and save it. You need to make sure your asset bundle is uncompressed by passing the `BuildAssetBundleOptions.UncompressedAssetBundle` argument here.

```csharp
using UnityEditor;
using UnityEngine;
using System.IO;

public class CreateAssetBundles
{
    [MenuItem("Assets/Build AssetBundles")]
    static void BuildAllAssetBundles()
    {
        string assetBundleDirectory = "Assets";

        BuildPipeline.BuildAssetBundles(assetBundleDirectory, BuildAssetBundleOptions.UncompressedAssetBundle, EditorUserBuildSettings.activeBuildTarget);
    }
}
```

7. Back in the Unity editor, select your font in the asset browser and include it in a AssetBundle by clicking the drop down box on the bottom of the inspector and clicking New. Name it whatever you want.

![4](https://github.com/user-attachments/assets/c87b1cd5-e6b2-47f4-9a4b-6c6311e2beec)

8. Go to the Assets menu and then click Build AssetBundles. **This option will only exist if you created the script from earlier correctly.** 

![5](https://github.com/user-attachments/assets/b95d85cd-cf88-4b6b-9918-a3f08b99c5c8)

9. Right click on the asset browser window and click Refresh to see the newly generated files. The only one you need is the one you created the name for earlier. In my case it was called `fontBundle`. Make sure it's just `fontBundle` and not `fontBundle.manifest`.

![6](https://github.com/user-attachments/assets/af946f21-3f38-4096-84d4-cc75eaf592d1)

![7](https://github.com/user-attachments/assets/20d276ee-036f-4ea1-a164-f878f495e2ef)

10. Copy the fontBundle file to 

```csharp

FF9_Data\EmbeddedAsset\FA

```

11. You can now load the font in Memoria by using `AssetBundle.CreateFromFile(string path)` and passing in the path to the `fontBundle` file.

12. You can modify the `LoadFont()` method within `EncryptFontManager` to use your font as the default font in game.  

```csharp

var newFontBundle = AssetBundle.CreateFromFile(PATH_TO_FONT_BUNDLE);
EncryptFontManager.defaultFont = EncryptFontManager.LoadFont(newFontBundle.LoadAsset<Font>("alpha brights"));

```

You need to make sure the string passed to `LoadAsset()` matches your fonts filename. You can check all the names of the files in your bundle by using `newFontBundle.GetAllAssetNames()`. It will return an array of strings that you an iterate through and log if you need. 

If all goes well you should see your font being used in-game.

![final](https://github.com/user-attachments/assets/9f52d883-fe72-4481-ab5b-6057e6f88662)

13. If you want to overwrite the `p_fa.mpc` file that comes with Memoria and FF9 then you need to encrypt your asset bundle. You can do this by using the `ByteEncryption` class like this:

```csharp

Byte[] bundleBytes = AssetManager.LoadBytes("EmbeddedAsset/FA/fontBundle");
Byte[] encrypted = ByteEncryption.Encryption(bundleBytes);

System.IO.File.WriteAllBytes(PATH_TO_WRITE_FILE_TO, encrypted);
```


If you put this code in the `LoadFont()` method then it will save the file when you start the game.

14. Make sure the file is named `p_fa.mpc` and is in the `FF9_Data\EmbeddedAsset\FA` directory. You then need to modify the `Memoria.ini` file to include the name of your font.

**Comment by Tirlititi** [comment]
Thanks for this. It is interesting for generating ``.mpc`` files with custom fonts embedded inside.
However, "in order to import a font, it must be inside a Unity asset bundle" is an incorrect statement: you can use any custom font that is installed on your machine as soon as it's installed for all users.

1) Install the font (typically a file with extension ``.ttf`` or ``.otf``) for all users. If installing for all users isn't the default behaviour, you can still right-click on the font and there should be an option there "Install font for all users".
2) Run the game and close it. This step is required so FFIX updates the list of fonts it can use.
3) Done: the font should now appear in the list of fonts that can be selected on the launcher.

**Comment by andy013** [comment]
@Tirlititi Yes, thank you for clarifying this. I should have mentioned that this is for developers who want to send font files to others along with Memoria. If you just want to use a custom font for yourself, then you can follow the directions you gave.
